### DISCUSSION TITLE
Overlay images with transparency

### DISCUSSION THREAD

**Discussion by Biome21** [ImageMagick/ImageMagick#4763] [Help] [answered]
Hi all,
I want to merge 2 pngs together. Lets call them "pngA" and "pngB". Both pngs have different values of transparency in their alpha channel. For the final image ("pngC"); I want to have pngA overlaid on top of pngB for rgb channels but for the a channel, I want the transparency of pngA. This is the command im using :
`convert pngA.png pngB.png -compose dst-over -composite pngC.png`

With this command the RGB channel have the proper value but the alpha is not the one from "pngA". If I use this command :
`convert pngA.png pngB.png -compose dst -composite pngC.png`

Then pngC has the proper alpha (the one of png A) but I lose the colors (RGB) of pngB for the transparent pixels.
This is probaly easy but I cant figure out the settings. Im using ImageMagick 7.0.11-12 Q16 x86_64. Heres an example and thank you for your time
PngA :
![pngA](https://user-images.githubusercontent.com/47017071/151203107-9b968f3e-0d09-451e-b9aa-bf78e5982fdb.png)
PngB :
![pngB](https://user-images.githubusercontent.com/47017071/151203190-3bcd3d0b-615c-4d38-8d97-ef565326fdf1.png)
PngC :
![pngC](https://user-images.githubusercontent.com/47017071/151203357-cbb275ea-b3d0-46ae-b4b1-b57cf534dbc3.png)

**Comment by fmw42** [comment]
This should do what you want, if I understand your request --  Unix syntax:

```
convert A.png B.png \
\( -clone 0 -alpha extract +write mpr:alpha +delete \) \
-alpha off -compose dst_over -composite \
mpr:alpha -alpha off -compose copy_opacity -composite \
AB.png

```

**Reply by Biome21** [reply]
Not quite. If I apply your command. The result has the proper alpha values (the one from A) but the transparent pixels dont have the rgb of B. Basically what im trying to achieve is given a png (A in this example) I want to pad the transparent pixels around the image with rgb value. Right now im trying to do this technique (https://legacy.imagemagick.org/Usage/canvas/#sparse_blur). Im blurring multiple times A with different sigma values then I merge all of these blur images into B. The final thing I want to do is have A (unblur) on top of B (blur) and have the transparent or semi transparent pixels of A have the rgb value of B. Maybe theres a simpler approach but im really close.

**Reply by fmw42** [reply]
You said overlay A onto B for the RGB values. That is what I did?  Do you want to overlay with transparency or just the RGB values?  Change dst_over to dst, if that is what you want. Also do not do the -alpha off before the -compose dst_over or dst as desire.  Sorry, but I took your explanation verbatim and implemented it. If you need something else, please explain better.

**Reply by fmw42** [reply]
> The final thing I want to do is have A (unblur) on top of B (blur) and have the transparent or semi transparent pixels of A have the rgb value of B. Maybe theres a simpler approach but im really close.

If you want A on top of B with B showing through the transparent pixels of A, then either

`convert A.png \( B.png -alpha off \) -compose dst_over -composite AB2.png `

![AB2](https://user-images.githubusercontent.com/13458218/151221014-560c9e9f-f318-4934-a595-72a24d4e555b.png)

or

```
convert A.png \
\( +clone -alpha extract -write mpr:alpha +delete \) \
\( B.png -alpha off \) -compose dst_over -composite \
mpr:alpha -alpha off -compose copy_opacity -composite \
AB3.png 
```

![AB3](https://user-images.githubusercontent.com/13458218/151221035-fc48a1b5-b85e-4df4-8894-e9f1865c1866.png)

**Reply by Biome21** [reply]
yes I want to overlay A on top of B. By overlaying A on top of B, it will :
- keep rgb value of A meaning that the opaque pixel of the result will have the value of A so the image will be unblur
- Keep the alpha of A. So the transparency is the same as A.

But for the transparent pixels of A, their rgb value should be the one from B which is not the case with your command or using "dest". Its not the case in the example images but if the transparent/semi transparent pixels have no rgb values (0) then those pixels in the final result should have the rgb of B (currently they stay black). Let me know if its not clear thanks

**Reply by fmw42** [reply]
Your images A and B are both blurred!?

**Reply by fmw42** [reply]
My commands do what you show in the two bullets. I  am not following your following paragraph.

" if the transparent/semi transparent pixels have no rgb values (0) then those pixels in the final result should have the rgb of B"

That is not clear to me.

**Reply by Biome21** [reply]
so the 2nd command you sent gives me what I want :
`convert A.png \
\( +clone -alpha extract -write mpr:alpha +delete \) \
\( B.png -alpha off \) -compose dst_over -composite \
mpr:alpha -alpha off -compose copy_opacity -composite \
AB3.png `

Now if you can give me a small explanation of what the arguments are doing (im gonna have to translate this into imagemagick.net later). It would help me alot understand my mistake.

For the clarification what I meant is even tho im overlaying A on top of B, if the transparent pixel or A dont have any RGB (0,0,0), the rgb in the result should have the value of B

**Reply by fmw42** [reply]
```
convert A.png \
\( +clone -alpha extract -write mpr:alpha +delete \) \
\( B.png -alpha off \) -compose dst_over -composite \
mpr:alpha -alpha off -compose copy_opacity -composite \
AB3.png 
```

```
Line 1:  read A.png
Line 2: copy A.png, extract its alpha channel and save to mpr: in-memory format and delete the copy
Line 3: read B.png and turn it's alpha channel off so only RGB channels, then compose A (RGBA) over B (RGB) using A's alpha channel to make B show through as needed
Line 4: read the mpr: image back, turn off alpha everywhere and put that alpha into the previously composited A over B
Line 5: save the result

```

**Reply by fmw42** [reply]
For Windows (DOS, not NET), the command would be:

```
convert A.png ^
( +clone -alpha extract -write mpr:alpha +delete ) ^
( B.png -alpha off ) -compose dst_over -composite ^
mpr:alpha -alpha off -compose copy_opacity -composite ^
AB3.png 
```

**Reply by Biome21** [reply]
Thank you very much for your help

**Reply by Biome21** [reply]
Hey @fmw42 me again. Im trying to convert all this into Magick.Net and ran into an issue. I can also repro it in command line so it might be easier. Do you know why with this command the transparent pixels of the result dont have any colors ?
`magick Color.png Mask.png -compose copy_opacity -composite Result.png `
Color.png has no transparency and is full of color. Mask.png only has alpha. I thought the command would keep the colors of Color.png and only modify the alpha channel according to the mask but it seems that the mask also erase the color value.

Thanks again

**Reply by snibgo** [reply]
> Mask.png only has alpha.

PNG files cannot contain just an alpha channel.

I suggest you supply sample Color.png and Mask.png files.

**Reply by Biome21** [reply]
I mean the other channels are empty so its black with different alpha values. What I would like is to use the rgb values of Color and the alpha of mask so the transparent pixels of result have a color. Heres an example:
Color.png:
![Color](https://user-images.githubusercontent.com/47017071/151448451-3160adc5-62d7-4f1e-8aed-d9abdb1f3433.png)
Mask.png:
![Mask](https://user-images.githubusercontent.com/47017071/151449238-81c67903-f4bd-4958-aba4-3d14c05937d5.png)
Result :
![Result](https://user-images.githubusercontent.com/47017071/151448455-aec87f4d-19a2-46ba-a39a-841bf1f685a3.png)

**Reply by snibgo** [reply]
Isn't the result what you want? If you remove alpha from that image it looks the same as Color.png.

**Reply by Biome21** [reply]
Heres an example of what I would like. The shape is not the same but you get the idea. If you compare both files in an image editing software youll see that the transparent pixels in my result wanted have a blue rgb where the result I posted the transparent pixels are black. Basically I dont want the mask to mess up the rgb values
![ResultWanted](https://user-images.githubusercontent.com/47017071/151451145-1d775b9f-3932-4255-8e7f-fd45d92ffad3.png)

**Reply by snibgo** [reply]
> ...where the result I posted the transparent pixels are black.

No. In the result you posted, the transparent pixels are mostly blue. We can show this by removing the alpha:
```
magick result.png -alpha off r2.png
```
r2.png is:
![r2](https://user-images.githubusercontent.com/33812368/151455014-05fc3872-2e0c-4c82-a955-958d6033f728.png)

**Reply by Biome21** [reply]
this is screenshots taken in gimp and turning off the alpha channel so we can see only rgb :
wanted :
![wanted](https://user-images.githubusercontent.com/47017071/151455475-4334c563-9f7a-43df-a833-136ac5f6d4fb.png)
result :
![result](https://user-images.githubusercontent.com/47017071/151455504-ec1c5f48-e5f2-4f3f-be25-d7f113d7d28d.png)

There is a difference. The wanted result png was also produce with a series of image magick command (posted by fred above) so I know theres a way to get there. Using image magick to turn off the alpha is not a valid test for me. Looking at it in photoshop or gimp or a game engine (since these pngs will be used in a game) is the real thing

**Reply by snibgo** [reply]
With Gimp 2.10.24 (revision 3), opening result.png, then menu Layer, Transparency, Remove alpha channel, I get a blue image.

In this image, IM is not changing the colour in the pixels that become transparent.

If turning the alpha off in Gimp shows black for you, then that is a problem with your Gimp. If your Gimp is older than mine, then try an upgrade. If your Gimp is newer than mine, then there has been a change in Gimp, perhaps a bug.

**Reply by Biome21** [reply]
So this is pretty weird, I did the same thing you did with 2 different versions of gimp and I get 2 different results (not the same as you)
In version 2.10.22 (revision 3) turning off the alpha shows black
![Screen Shot 2022-01-27 at 9 09 26 PM](https://user-images.githubusercontent.com/47017071/151552882-30cd1353-5622-424a-aea2-19ff0bf0b79b.png)

In version 2.10.20 (latest) I get white
![Screen Shot 2022-01-28 at 8 14 43 AM](https://user-images.githubusercontent.com/47017071/151553060-1fd82d3f-074f-49fa-90ad-e1cfd4021a47.png)

Both of the results is not what I want (blue) but these pngs will be in unity and unity pick up the right rgb value so I will keep my solution.

Sorry for the time wasted but at the same time its pretty crazy that gimp has a bug this big.

**Reply by snibgo** [reply]
I suggest you raise this on a Gimp forum, or report this as a bug to Gimp.

**Comment by Biome21** [comment, marked as answer]
this was the answer from @fmw42 that worked for me :
convert A.png \
\( +clone -alpha extract -write mpr:alpha +delete \) \
\( B.png -alpha off \) -compose dst_over -composite \
mpr:alpha -alpha off -compose copy_opacity -composite \
AB3.png
